### PR TITLE
Fix #37, Sends message when file complete

### LINKED ISSUE #37: Send a message when new files are completed
**Checklist (Please check before submitting)**

* [x] I reviewed the [Contributing Guide](https://github.com/nasa/DS/blob/main/CONTRIBUTING.md).
* [x] I reviewed the [README file](https://github.com/nasa/DS/blob/main/README.md) to see if the feature is in the major future work.
* [x] I performed a cursory search to see if the feature request is relevant, not redundant, nor in conflict with other tickets.

**Is your feature request related to a problem? Please describe.**
Send a message when new files are completed in DS.

**Describe the solution you'd like**

- Initialize and send a message using `CFE_MSG_INIT()` and `CFS_SB_TransmitMsg()` respectively. 
- Define a new packet `DS_FileCompletePkt_t` for a single `DS_FileInfo_t` payload.
- Define a new MsgID.
- May need to include logic that will use "movename" when DS_MOVE_FILES == FALSE and use "filename" when DS_MOVES_FILES == TRUE. 
- This feature will be implemented at the close of a file within the `FileCloseDest()` function.

**Describe alternatives you've considered**
None

**Additional context**
None

**Requester Info**
Justin Figueroa, ASRC Federal

### PR DESCRIPTION
**Pull request by chillfig**
**Checklist (Please check before submitting)**

* [x] I reviewed the [Contributing Guide](https://github.com/nasa/DS/blob/main/CONTRIBUTING.md).
* [x] I signed and emailed the appropriate [Contributor License Agreement](https://github.com/nasa/cFS/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla) to GSFC-SoftwareRelease@mail.nasa.gov and copied cfs-program@lists.nasa.gov.

**Describe the contribution**
A clear and concise description of what the contribution is.
- Fixes #37

**Testing performed**
Steps taken to test the contribution:
1. Build steps '...'
1. Execution steps '...'

**Expected behavior changes**
Generate a message when a file gets completed.

**System(s) tested on**
 - Ubuntu 18.04

**Additional context**
None

**Third party code**
If included, identify any third party code and provide text file of license

**Contributor Info - All information REQUIRED for consideration of pull request**
Justin Figueroa, ASRC Federal

### CODE REVIEW

**Comment by chillfig** [comment]
@skliper Do I manually implement the sub function in apps/ds/unit-test/stubs in this PR?

**Review by skliper — COMMENTED** [review]
Great start!  See comments.  Since the new function is only called from ds_file.c it doesn't seem like you need to implement a stub.  Just add a unit test within the ds_file_tests.c.  Should also consider adding a requirement and updating the design slides, although the design slides aren't configuration managed within the repo anymore so we'll need to establish a new process.

File: fsw/src/ds_file.c [hunk 1]
```diff
@@ -1025,6 +1031,31 @@ void DS_FileTestAge(uint32 ElapsedSeconds)
...
 
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+/*                                                                 */
+/* DS_FileTransmit() - transmit file info                          */
+/*                                                                 */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+void DS_FileTransmit(DS_AppFileStatus_t *FileStatus)
+{
+    DS_FileCompletePkt_t DS_FileCompletePkt = {.FileStatus = *FileStatus};
```
Comment (skliper) [inline fsw/src/ds_file.c:1042]: Recommend using the FileInfo structure in the packet and copy elements from FileStatus into FileInfo.

File: fsw/src/ds_msg.h [hunk 2]
```diff
@@ -356,6 +357,16 @@ typedef struct
...
 } DS_FileInfoPkt_t;
 
+/**
+ * \brief Completed application file info
+ */
+typedef struct
+{
+    CFE_MSG_TelemetryHeader_t TlmHeader; /**< \brief cFE Software Bus telemetry message header */
+
+    DS_AppFileStatus_t FileStatus; /**< \brief Current state of destination file */
```
Comment (skliper) [inline fsw/src/ds_msg.h:367]: Make this a single DS_FileInfo_t element

**Review by skliper — CHANGES_REQUESTED** [review]

File: fsw/src/ds_file.c [hunk 3]
```diff
@@ -1025,6 +1036,61 @@ void DS_FileTestAge(uint32 ElapsedSeconds)
...
+    */
+    DS_FileCompletePkt.FileInfo.SequenceCount = FileStatus->FileCount;
+    /*
+    ** Set file enable state...
+    */
+    DS_FileCompletePkt.FileInfo.EnableState = FileStatus->FileState;
+    /*
+    ** Set file open state...
+    */
+    DS_FileCompletePkt.FileInfo.OpenState = DS_OPEN;
```
Comment (skliper) [inline fsw/src/ds_file.c:1078]: I think this should be DS_CLOSE?

File: fsw/src/ds_file.c [hunk 4]
```diff
@@ -1025,6 +1036,63 @@ void DS_FileTestAge(uint32 ElapsedSeconds)
...
 
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+/*                                                                 */
+/* DS_FileTransmit() - transmit file info                          */
+/*                                                                 */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+void DS_FileTransmit(DS_AppFileStatus_t *FileStatus)
+{
+    DS_FileCompletePkt_t DS_FileCompletePkt;
```
Comment (skliper) [inline fsw/src/ds_file.c:1047]: I think this actually needs to be updated... should be a pointer.

File: fsw/src/ds_msg.h [hunk 5]
```diff
@@ -356,6 +356,22 @@ typedef struct
...
+    CFE_MSG_TelemetryHeader_t TlmHeader; /**< \brief cFE Software Bus telemetry message header */
+
+    DS_FileInfo_t FileInfo; /**< \brief Current state of destination file */
+} DS_FileCompletePkt_t;
+
+typedef union
+{
+    CFE_SB_Buffer_t      SBBuf;
+    DS_FileCompletePkt_t Pkt;
+} DS_FileCompletePktBuf_t;
```
Comment (skliper) [inline fsw/src/ds_msg.h:373]: Add doxygen documentation (this typedef supports CFE SB Allocate buffer use with a DS_FileCompletePkt_t that matches the alignment constraints of a SBBuf)

**Comment by skliper** [comment]
@dzbaker I think this is ready to merge.

**Pull request merged**